# DATA CLEANING PROJECT
**by Jayesh Suryawanshi**    
  
  
  
  **Information on the Data Source:**  
The data source consists of an two csv files namely fifa21_raw_data.csv and fifa21 raw data v2.csv which consists of information on football players as of EA Sports' FIFA 21 release. It was scraped from the website sofifa.com and contains information on players' biometrics, demographics, stats, etc. The dataset is available on Kaggle but unlike most datasets available on the website, this dataset is highly unclean, messy and raw.  

Let us clean it up and make it so that it can be used to perform analysis tasks, etc.  

The tool we will use to do so is Python.

Let's start by importing the standard data cleaning and analysis libraries. We can import others later as and when required.

In [1]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
pd.options.display.max_columns = 200

Here, we will work with one of the two data files provided on Kaggle. This version is much more messy so we can put our data cleaning skills to the test.  
  
  Let's start by reading the data into a pandas dataframe.

In [2]:
df = pd.read_csv('fifa21_raw_data.csv')

C:\Users\Jayesh\AppData\Local\Temp\ipykernel_22592\830073151.py:1: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('fifa21_raw_data.csv')


In [3]:
df.shape

(18979, 77)

So, there are 77 columns and 18979 rows in this dataset. We can take a look at how the data actually looks.

In [4]:
df.head()

photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                       LongName  \
0                  Lionel Messi   
1  C. Ronaldo dos Santos Aveiro   
2                     Jan Oblak   
3               Kevin De Bruyne   
4    Neymar da Silva Santos Jr.   

                                           playerUrl Nationality Positions  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina  RW ST CF   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal     ST LW   
2  http://sofifa.com/player/200389/jan-oblak/210005/    Slovenia        GK   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium    CAM CM   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil    LW CAM   

                Name  Age  ↓OVA  POT  \
0           L. Messi   33    93   93   
1  Cristiano Ronaldo   35    92   92   
2           J. Oblak   27    91   93   
3       K. De Bruyne   29    91   91   
4          Neymar Jr   28    91   91   

                                Team & Contract      ID Height  Weight   foot  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n  158023   5'7"  159lbs   Left   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   20801   6'2"  183lbs  Right   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n  200389   6'2"  192lbs  Right   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  192985  5'11"  154lbs  Right   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n  190871   5'9"  150lbs  Right   

   BOV   BP  Growth        Joined Loan Date End   Value   Wage Release Clause  \
0   93   RW       0   Jul 1, 2004           NaN  €67.5M  €560K        €138.4M   
1   92   ST       0  Jul 10, 2018           NaN    €46M  €220K         €75.9M   
2   91   GK       2  Jul 16, 2014           NaN    €75M  €125K        €159.4M   
3   91  CAM       0  Aug 30, 2015           NaN    €87M  €370K          €161M   
4   91   LW       0   Aug 3, 2017           NaN    €90M  €270K        €166.5M   

   Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  \
0        429        85         95                70             91       88   
1        437        84         95                90             82       86   
2         95        13         11                15             43       13   
3        407        94         82                55             94       82   
4        408        85         87                62             87       87   

   Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
0    470         96     93           94            91            96       451   
1    414         88     81           76            77            92       431   
2    109         12     13           14            40            30       307   
3    441         88     85           83            93            92       398   
4    448         95     88           89            81            95       453   

   Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  \
0            91            80       91         94       95    389          86   
1            87            91       87         95       71    444          94   
2            43            60       67         88       49    268          59   
3            77            76       78         91       76    408          91   
4            94            89       96         91       83    357          80   

   Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  \
0       68       72        69          94        347          44   
1       95       84        78          93        353          63   
2       78       41        78          12        140          34   
3       63       89        74          91        408          76   
4       62       81        5

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   photoUrl          18979 non-null  object
 1   LongName          18979 non-null  object
 2   playerUrl         18979 non-null  object
 3   Nationality       18979 non-null  object
 4   Positions         18979 non-null  object
 5   Name              18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Team & Contract   18979 non-null  object
 10  ID                18979 non-null  int64 
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  foot              18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  BP                18979 non-null  object
 16  Growth            18979 non-null  int64 
 17  Joined      

This step helps us get an overall look regarding the various attributes (columns) in the dataset. If you look closely, the column **Loan Date End** consists of a lot of null values. It makes sense as in football, only some players leave on loan to other clubs so the players who are not on loan will not have a loan end date and that explains the null values in that column. We will handle it later.

In [6]:
df.isna().sum()

photoUrl       0
LongName       0
playerUrl      0
Nationality    0
Positions      0
              ..
PAS            0
DRI            0
DEF            0
PHY            0
Hits           0
Length: 77, dtype: int64

In [7]:
df[~df.isna()].shape

(18979, 77)

There are no other columns with null values.

It can be seen that there are two columns that specify the player names, however one of the columns specifies it in short and it isn't much helpful. There is another column called growth which specifies the player's rating growth from previous season (previous fifa version) in numbers. It isn't significant either. So these two columns can be dropped.

In [8]:
df.drop(['Name','Growth'],axis='columns',inplace=True)

Next, let's check out the weight column to see if there are any inconsistent units.

In [9]:
print("There are {num} rows containing \"kg\".".format(num=df.Weight[df.Weight.str.contains(r"kg")].shape[0]))
print("There are {num} rows containing \"lbs\".".format(num=df.Weight[df.Weight.str.contains(r"lbs")].shape[0]))

There are 0 rows containing "kg".
There are 18979 rows containing "lbs".


All weight units are specified in lbs.

In [10]:
# df['Weight'].unique()

Similarly, let's check if there are any inconsistencies in the Height column.

In [11]:
print("There are {num} rows containing \"cm\".".format(num=df.Height[df.Height.str.contains(r"cm")].shape[0]))

There are 0 rows containing "cm".


We can now make widespread changes to multiple columns and convert them to the readable format we require. This can be done using the pandas apply function. Before doing that, let's take a look at the columns and the unique values they consist of so that we can spot any odd values that don't match with the existing format. 

In [12]:
df['Height'].unique()

array(['5\'7"', '6\'2"', '5\'11"', '5\'9"', '6\'0"', '5\'10"', '6\'3"',
       '6\'4"', '6\'1"', '6\'6"', '5\'8"', '5\'6"', '6\'5"', '5\'5"',
       '5\'4"', '6\'7"', '5\'2"', '5\'3"', '6\'8"', '5\'1"', '6\'9"'],
      dtype=object)

All values in the Height column are of the format - Height in ft and inches.

Let's convert all the values in the **Weight** column from lbs to kgs and all the values from the **Height** column to centimeters so that they can be represented numerically and not as text.

The **Value** column signifies the player's market value. On reviewing the dataframe, it can be seen that the values are accompanied with some abbreviations. Let's find such unique abbreviations.

In [13]:
df['Value'].str[-1].unique()

array(['M', '0', 'K'], dtype=object)

The unique abbreviations are 'K' and 'M' which means the player values are in thousands or millions. There are also some players whose value is not represented in either of the abbreviations but plain numbers.

Similarly, the **Wage** column represents the players wages.

In [14]:
df['Wage'].str[-1].unique()

array(['K', '0'], dtype=object)

We can convert the values with 'M' and 'K' abbevations to numeric by multiplying them with 1,000,000 and 1,000 respectively.

The W/F column represents the player's weak foot skills in stars. We find out the unique star ratings assigned to player's weak foots'.

In [15]:
df['W/F'].unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

Similarly, SM refers to Skill Moves.

In [16]:
df['SM'].unique()

array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

IR refers to International Reputation.

In [17]:
df['IR'].unique()

array(['5 ★', '3 ★', '4 ★', '2 ★', '1 ★'], dtype=object)

So, for the W/F, SM and IR columns, we can remove the ★ and just represent the values with integers in the range [1,5]

Now, that we have an idea of what changes we need to make to each of the columns discussed above, we can write a function for the same and apply it to the entire dataframe.

In [18]:
def changes(row):
    a = row['Height']
    a="".join(ch for ch in a if ch.isnumeric()) # strip of the string to extract numerical values and remove the ' and " symbols
    row['Height'] = (int(a[0])*12+int(a[1:]))*2.54 # Using type conversion and formulas to convert the extracted numerical values for feet and inches to centimeters
    
    b = row['Weight']
    row['Weight'] = round(int("".join(ch for ch in b if ch.isnumeric()))*0.45359237,2) # extracting numericals values from the string, converting it to kgs and finally rounding it up to 2 decimals

    c = row['Value']
    if c[-1] == 'M' or c[-1] == 'm':
       row['Value'] = int("".join(ch for ch in c if ch.isnumeric()))*1000000 # extracting numericals values from the string, typecasting it to integer and multiplying it with 1000000 if the abbreviation it accompanied was 'M'
    elif c[-1] == 'K' or c[-1] == 'k':
       row['Value'] = int("".join(ch for ch in c if ch.isnumeric()))*1000 # extracting numericals values from the string, typecasting it to integer and multiplying it with 1000 if the abbreviation it accompanied was 'K'
    else:
       row['Value'] = int("".join(ch for ch in c if ch.isnumeric())) # if the player value had no abbreviation, then it indicates that it consisted of plain numbers, so we strip of any spaces or special characters and typecast it to integer

    d = row['Wage']
    if d[-1] == 'M' or d[-1] == 'm':
       row['Wage'] = int("".join(ch for ch in d if ch.isnumeric()))*1000000 # extracting numericals values from the string, typecasting it to integer and multiplying it with 1000000 if the abbreviation it accompanied was 'M'
    elif d[-1] == 'K' or d[-1] == 'k':
       row['Wage'] = int("".join(ch for ch in d if ch.isnumeric()))*1000 # extracting numericals values from the string, typecasting it to integer and multiplying it with 1000 if the abbreviation it accompanied was 'K'
    else:
       row['Wage'] = int("".join(ch for ch in d if ch.isnumeric())) # if the player wages had no abbreviation, then it indicates that it consisted of plain numbers, so we strip of any spaces or special characters and typecast it to integer

    e = row['Release Clause']
    if e[-1] == 'M' or e[-1] == 'm':
       row['Release Clause'] = int("".join(ch for ch in e if ch.isnumeric()))*1000000 # extracting numericals values from the string, typecasting it to integer and multiplying it with 1000000 if the abbreviation it accompanied was 'M'
    elif e[-1] == 'K' or e[-1] == 'k':
       row['Release Clause'] = int("".join(ch for ch in e if ch.isnumeric()))*1000 # extracting numericals values from the string, typecasting it to integer and multiplying it with 1000 if the abbreviation it accompanied was 'K'
    else:
       row['Release Clause'] = int("".join(ch for ch in e if ch.isnumeric())) # if the player value had no abbreviation, then it indicates that it consisted of plain numbers, so we strip of any spaces or special characters and typecast it to integer
   
    f = row['W/F']
    row['W/F'] = int("".join(ch for ch in f if ch.isnumeric())) # stripping the stars and converting the rating to integer

    g = row['SM']
    row['SM'] = int("".join(ch for ch in g if ch.isnumeric())) # stripping the stars and converting the rating to integer
    
    h = row['IR']
    row['IR'] = int("".join(ch for ch in h if ch.isnumeric())) # stripping the stars and converting the rating to integer
    
    return row

df = df.apply(changes,axis='columns')

Now that the required changes have been applied and the columns are converted to the desired format, we rename the columns as some column names are represented using the shortforms of the actual names.

In [19]:
df.rename(columns={'Height':'Height (cm)','Weight':'Weight (kg)','Value':'Value (Euros)','Wage':'Wages (Euros)','Release Clause':'Release Clause (Euros)','W/F':'Weak Foot','SM':'Skill Moves','IR':'International Reputation','A/W':'Attack Workrate','D/W':'Defense Workrate','BOV':'Highest Rating','BP':'Best Position','POT':'Potential','↓OVA':'Rating','foot':'Strong foot','LongName':'Name'},inplace=True)

Now, there's a column **Team & Contract** which consists of the players' club name and contract validity. Let's split it into two separate columns and remove the unnecessary 'newline' characters.

In [20]:
# creating two new columns and initially assigning them the same values as the Team & Contract column
df['Club'] = df['Team & Contract']  
df['Contract'] = df['Team & Contract']
def changes(row):
    a = row['Team & Contract'].split('\n') # splitting the team and contract column into a list based on the newline character
    ele=''
    x=[i for i in a if i!=ele]
    row['Club'] = x[0] # assigning the first element of the list to the club column
    row['Contract'] = x[1]  # assigning the second element to the contract column

    
    return row

df = df.apply(changes,axis='columns')
df.drop(['Team & Contract'], axis='columns',inplace=True) # The Team & Contract column is no longer required so we just go ahead and drop it.
    
    

Next let's check if there are any Clubs with numbers or other special charcters in their name.

In [21]:
df.Club[df.Club.str.contains(r"^\d+")].value_counts()

1. FSV Mainz 05          32
1. FC Köln               31
1. FC Nürnberg           30
1. FC Heidenheim 1846    30
1. FC Union Berlin       28
1. FC Kaiserslautern     28
1. FC Magdeburg          28
1. FC Saarbrücken        27
Name: Club, dtype: int64

It can be seen that there are some clubs whose names are starting with '1.'. We should address this problem.

In [22]:
df['Club'] = df['Club'].map(lambda x: x.replace('1. ',''))

In [23]:
df.Club[df.Club.str.contains(r"^\d+")].value_counts()

Series([], Name: Club, dtype: int64)

Next let's take a look at the unique contract lengths and periods

In [24]:
df['Contract'].value_counts().head()

2019 ~ 2021    1706
2020 ~ 2022    1445
2020 ~ 2021    1440
2019 ~ 2022    1236
2018 ~ 2021    1163
Name: Contract, dtype: int64

Now that we have an idea about what the Contract column looks like, let's get a look at whether there are other values in the column that don't correspond to the 'YEAR ~ YEAR' format.

In [25]:
df['Contract'][~df['Contract'].str.contains('~')].value_counts()

Jun 30, 2021 On Loan    770
Free                    238
Dec 31, 2020 On Loan    100
May 31, 2021 On Loan     30
Dec 31, 2021 On Loan     22
Jan 31, 2021 On Loan     18
Jan 1, 2021 On Loan      15
Jun 30, 2022 On Loan     15
Nov 30, 2020 On Loan      7
Jun 23, 2021 On Loan      7
Nov 27, 2021 On Loan      6
Nov 22, 2020 On Loan      5
Jan 3, 2021 On Loan       3
Jan 4, 2021 On Loan       2
Aug 31, 2021 On Loan      2
May 31, 2022 On Loan      2
Dec 30, 2020 On Loan      1
Jul 31, 2021 On Loan      1
Dec 30, 2021 On Loan      1
Jun 30, 2023 On Loan      1
Jan 17, 2021 On Loan      1
Jul 1, 2021 On Loan       1
Jul 5, 2021 On Loan       1
Jan 30, 2021 On Loan      1
Aug 1, 2021 On Loan       1
Name: Contract, dtype: int64

It can be seen that there are infact Contract end dates that don't match the format 'YEAR ~ YEAR', and they all belong to players on loan. There are also players who are under no contract that are represented using 'Free'.

Let's make some changes and simplify things a bit. We will replace ' ~ ' with '-', 'Free' with 'No contract' and '(Date) On Loan' with 'Till (Date)'.

In [26]:
df['Contract'] = df['Contract'].map(lambda x:x.replace(' ~ ','-').replace(' On Loan',''))
df['Contract'] = df['Contract'].map(lambda x: x if '-' in x or 'Free' in x else 'Till ' + x)
df['Contract'] = df['Contract'].map(lambda x: 'No contract' if 'Free' in x else x)

Let's take a look at players who are under no contract. As they are under no contract, they should belong to no clubs. Let's check out what value is assigned under the Club column to such players.

In [27]:
df[df['Contract']=='No contract']['Club'].unique()

array([' Brazil', ' Ecuador', ' Uruguay', ' Colombia', ' Russia',
       ' Mexico', ' Peru', ' Sweden', ' Paraguay', ' Australia', ' Egypt',
       ' Canada', ' Norway', ' South Africa', ' Hungary', ' Slovenia',
       ' Ivory Coast', ' Bulgaria', ' Iceland', ' Greece', ' Poland',
       ' Venezuela', ' Bolivia', ' Romania', ' Northern Ireland',
       ' Wales', ' Chile', ' Finland', ' Cameroon', ' China PR', ' India',
       ' New Zealand', ' United Arab Emirates'], dtype=object)

It is seen that the players who have no contract, have their country names mentioned in the clubs column. We can take care of this by assigning the value 'No club' to the Club column for all players under no contract.

In [28]:
def changes(row):
    if row['Contract'] == 'No contract':
        row['Club'] = 'No club'
    return row
df = df.apply(changes,axis='columns')

# df2['Club'] = df2['Club'].apply(lambda x: 'No club' if 'No contract' in df2['Contract'] else x) can also be used

In [29]:
df[df['Contract']=='No contract']['Club'].unique()

array(['No club'], dtype=object)

Players can play at different positions. In the **Positions** column, the various positions for each player are separated using spaces. It makes it difficult to read and interpret. We assign a proper separator so that positions become readable.

In [30]:
df['Positions'] = df['Positions'].map(lambda x:x.replace(' ',', '))

When I opened one of the photo URLs, the sites were irresponsive. After a bit of research this is due to a domain name change in the site from where the data is scraped. We thus need to change the url format and can do so for each URL using the following code.

In [31]:
df['photoUrl'] = df['photoUrl'].map(lambda x:x.replace('60.png','120.png').replace('.com','.net'))
# df['photoUrl'] = df['photoUrl'].map(lambda x:x.replace('.com','.net')) can also be used

There is another column **Hits** with newline characters and other abbreviations. It will be better represented as a numeric column.

In [32]:
df.Hits.unique()

array(['\n372', '\n344', '\n86', '\n163', '\n273', '\n182', '\n646',
       '\n79', '\n164', '\n170', '\n93', '\n131', '\n89', '\n90', '\n169',
       '\n187', '\n103', '\n317', '\n239', '\n56', '\n130', '\n229',
       '\n124', '\n172', '\n114', '\n46', '\n663', '\n479', '\n203',
       '\n198', '\n122', '\n565', '\n152', '\n140', '\n276', '\n209',
       '\n25', '\n136', '\n109', '\n139', '\n42', '\n99', '\n137',
       '\n166', '\n80', '\n104', '\n150', '\n112', '\n285', '\n133',
       '\n96', '\n69', '\n143', '\n100', '\n38', '\n83', '\n67', '\n207',
       '\n416', '\n315', '\n206', '\n424', '\n191', '\n454', '\n271',
       '\n249', '\n121', '\n246', '\n171', '\n148', '\n329', '\n70',
       '\n97', '\n73', '\n210', '\n82', '\n146', '\n62', '\n165', '\n129',
       '\n162', '\n35', '\n65', '\n47', '\n77', '\n106', '\n84', '\n1.3K',
       '\n310', '\n181', '\n364', '\n200', '\n141', '\n134', '\n259',
       '\n309', '\n367', '\n224', '\n211', '\n60', '\n176', '\n108',
       '\n

The above code shows that most values are integers represented as text, however there are other such as 2.9K and 1.3K which should be handled. There are some records with actual integer values. This means that this column is highly inconsistent.

In [33]:
def changes(row):
    a = str(row['Hits']).split('\n')    # first we convert the values to integer (since there are some int values as well) and split the string based on the new line character and create a list
    x = [i for i in a if i!=''] # next we remove all the null characters
    x = ''.join(x) 
    if x[-1] == 'K' or x[-1] == 'k':
        row['Hits'] = int(float(x[:-1])*1000) # if the value is accompanied with an abbreviation K or k then we multiply it with 1000
    else:
        row['Hits'] = int(x)
    return row
    
df = df.apply(changes,axis='columns')

Now that most of the required columns are cleaned, let's create a column which will help us understand how long a player has been at the club

In [34]:
df['Joined2'] = df['Joined']
df['Joined2'] = df['Joined2'].astype('datetime64')
df['Joined2'] = pd.to_datetime(df['Joined2'])

In [35]:
from datetime import datetime # we import the require datetime library

date_to_subtract = datetime.strptime('2020-08-14', '%Y-%m-%d').date()   #release date of fifa 21

df['Club Duration'] = date_to_subtract - df['Joined2'].dt.date #indicates how long the player has been at the club with respect to the fifa21 release date

def changes(row):
    if row['Club'] == 'No club':
        row['Club Duration'] = 'Not Available'
    else:
        a = int(str(row['Club Duration']).split(' ')[0])
        years = a//365
        months = (a%365)//30
        days = (a%365)%30
        if years==1:
            y = 'Year'
        else:
            y='Years'
        if months == 1:
            m = 'Month'
        else:
            m = 'Months'
        if days == 1:
            d = 'Day'
        else:
            d = 'Days'
        row['Club Duration'] = f'{years} {y} {months} {m} {days} {d}'
    return row
df = df.apply(changes,axis='columns')
df.drop(['Joined2'],axis='columns',inplace=True) # we drop the column as it is no longer required

Here, we assigned 'Not Available' as the value to club duration for players who are under no contract.

We can also create a column which indicates whether a player is on Loan or not

In [36]:
df['On Loan'] = np.where(df['Loan Date End'].isnull(), 'No', 'Yes')

The Loan Date End Column no longer plays any significant role and can be dropped.

In [37]:
df.drop(['Loan Date End'],axis='columns',inplace=True)

Next, we rearrange all the columns in order of priority.

In [38]:
columns = ['ID','Name', 'Nationality', 'Age', 'Height (cm)', 'Weight (kg)', 
       'Club', 'Contract', 'Joined',
       'Club Duration', 'On Loan','Rating', 'Potential',  
        'Highest Rating', 'Positions', 'Best Position', 
       'Value (Euros)', 'Wages (Euros)', 'Release Clause (Euros)', 'Strong foot', 'Attacking',
       'Crossing', 'Finishing', 'Heading Accuracy', 'Short Passing', 'Volleys',
       'Skill', 'Dribbling', 'Curve', 'FK Accuracy', 'Long Passing',
       'Ball Control', 'Movement', 'Acceleration', 'Sprint Speed', 'Agility',
       'Reactions', 'Balance', 'Power', 'Shot Power', 'Jumping', 'Stamina',
       'Strength', 'Long Shots', 'Mentality', 'Aggression', 'Interceptions',
       'Positioning', 'Vision', 'Penalties', 'Composure', 'Defending',
       'Marking', 'Standing Tackle', 'Sliding Tackle', 'Goalkeeping',
       'GK Diving', 'GK Handling', 'GK Kicking', 'GK Positioning',
       'GK Reflexes', 'Total Stats', 'Base Stats', 'Weak Foot', 'Skill Moves',
       'Attack Workrate', 'Defense Workrate', 'International Reputation',
       'PAC', 'SHO', 'PAS', 'DRI', 'DEF', 'PHY', 'Hits', 'photoUrl','playerUrl']
df = df.reindex(columns,axis='columns')

In [39]:
df.head()

ID                          Name Nationality  Age  Height (cm)  \
0  158023                  Lionel Messi   Argentina   33       170.18   
1   20801  C. Ronaldo dos Santos Aveiro    Portugal   35       187.96   
2  200389                     Jan Oblak    Slovenia   27       187.96   
3  192985               Kevin De Bruyne     Belgium   29       180.34   
4  190871    Neymar da Silva Santos Jr.      Brazil   28       175.26   

   Weight (kg)                 Club   Contract        Joined  \
0        72.12         FC Barcelona  2004-2021   Jul 1, 2004   
1        83.01             Juventus  2018-2022  Jul 10, 2018   
2        87.09      Atlético Madrid  2014-2023  Jul 16, 2014   
3        69.85      Manchester City  2015-2023  Aug 30, 2015   
4        68.04  Paris Saint-Germain  2017-2022   Aug 3, 2017   

               Club Duration On Loan  Rating  Potential  Highest Rating  \
0   16 Years 1 Month 18 Days      No      93         93              93   
1     2 Years 1 Month 6 Days      No      92         92              92   
2      6 Years 1 Month 1 Day      No      91         93              91   
3  4 Years 11 Months 21 Days      No      91         91              91   
4   3 Years 0 Months 12 Days      No      91         91              91   

    Positions Best Position  Value (Euros)  Wages (Euros)  \
0  RW, ST, CF            RW      675000000         560000   
1      ST, LW            ST       46000000         220000   
2          GK            GK       75000000         125000   
3     CAM, CM           CAM       87000000         370000   
4     LW, CAM            LW       90000000         270000   

   Release Clause (Euros) Strong foot  Attacking  Crossing  Finishing  \
0              1384000000        Left        429        85         95   
1               759000000       Right        437        84         95   
2              1594000000       Right         95        13         11   
3               161000000       Right        407        94         82   
4              1665000000       Right        408        85         87   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                70             91       88    470         96     93   
1                90             82       86    414         88     81   
2                15             43       13    109         12     13   
3                55             94       82    441         88     85   
4                62             87       87    448         95     88   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           94            91            96       451            91   
1           76            77            92       431            87   
2           14            40            30       307            43   
3           83            93            92       398            77   
4           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         91       32   
1           95      82         84 

Finally, we save the cleaned dataset to a csv file.

In [40]:
df.to_csv('cleaned_fifa_21_data.csv',index=False)